### DISCUSSION TITLE
Tutorial Python Code Error

### DISCUSSION THREAD

**Discussion by bytetec** [FLAMEGPU/FLAMEGPU2#1076] [Q&A] [answered]
Hi everyone

I tried the pure python tutorial code but it fails with "list index out of range" error.
![image](https://github.com/FLAMEGPU/FLAMEGPU2/assets/138150718/294f338c-6ba0-4671-b5f3-ff70c6008dbe)

https://docs.flamegpu.com/tutorial/index.html#tutorial-creating-the-circles-model

Has anybody an idea how to fix this?
Thx in advance for helping.

**Comment by Robadob** [comment]
Hi,

I'm not too familiar with the Python native API (It was developed by @mondus).

I just built `pyflamegpu` from `master` (with Python 3.9) on my local machine and tried to run the full tutorial source (from the end of [the page](https://docs.flamegpu.com/tutorial/index.html#complete-tutorial-code)).

There was an error, `addRoot()` at line 89 should be `addExecutionRoot()` (I'll correct this ASAP), but after fixing that the tutorial ran without an exception being thrown.

It's possible you've made a mistake in following the tutorial. Based on the specific code where the error is occurring, I would assume that method being passed to `pyflamegpu.codegen.translate()` is incorrectly defined.

Could you possibly share your code in a [gist](https://gist.github.com) so that I can investigate further (and perhaps add a clearer error to the codegen function).

**Comment by bytetec** [comment]
Hi Robadob
Code is an 1:1 copy from the tutorial page plus the addRoot() correction. 
I'm using Python 3.11 could that be an issue? Otherwise I think it is something with the c++ setup, I'm not very familiar with that.
![image](https://github.com/FLAMEGPU/FLAMEGPU2/assets/138150718/75e6b19f-8374-458e-8f7e-baf5b3ef6fe0)
This is all I get.

**Reply by Robadob** [reply]
The code your error is [coming from](https://github.com/FLAMEGPU/FLAMEGPU2/blob/master/swig/python/codegen/__init__.py) is all Python. The python native translation relies on core Python's own syntax parsing modules. It's possible this has changed with Python 3.11 leading to the error.

I will test it on my office machine which has the latest Python setup, but I'm unlikely to find a resolution until tomorrow (it's late evening here).

**Reply by bytetec** [reply]
Thank you, I'm not in a hurry :-)

**Reply by Robadob** [reply]
Built `pyflamegpu` from the [`bugfix_pynative_2023_07_05`](https://github.com/FLAMEGPU/FLAMEGPU2/pull/1077) branch (this just fixes the pynative translation bug that I noticed whilst investigating your first post) with Python 3.11.4 installed via miniconda, and it runs fine for me.

![image](https://github.com/FLAMEGPU/FLAMEGPU2/assets/742154/350f5673-5e48-4874-b5b7-be36e8b46ea8)

Could you perhaps download [this](https://drive.google.com/file/d/1M7CWqTAwY1QL1GahAK7ZDhPyHL4_t6CQ/view?usp=sharing) copy of the tutorial, the same version I'm running and check whether that works? My current best guess is that you've got some unusual/subtle difference in your copy somehow.

If that fails, next step would be having you run it through a standalone copy of `codegen.py` so we can actually log some information to catch where the difference is coming from.

**Reply by Robadob** [reply]
I also noticed in the second screenshot you posted, you're use PyCharm (and it's included in the callstack of the error). It's possible PyCharm is running the code weirdly to enable debugging and this interfering with how codegen grabs the code. Are you able to try and run the tutorial outside of pycharm? If that's confirmed to be the problem we can look into it and find a resolution.

**Reply by Robadob** [reply, marked as answer]
Ok I've thought about this overnight.

I'm fairly certain PyCharm's PyDev debugger's `exec(compile(contents+"\n", file, 'exec'), glob, loc)` is to blame. Somehow, it's breaking codegen's detection of the function.

I've tried moving the tutorial inside a module, but that doesn't break it. So I'll need to find time to install this module to debug this further.

A temporary workaround for yourself would be to instead pass the agent functions as strings, bypassing the detection that is failing.

The original fn

```py
@pyflamegpu.agent_function
def output_message(message_in: pyflamegpu.MessageNone, message_out: pyflamegpu.MessageSpatial2D):
    message_out.setVariableUInt("id", pyflamegpu.getID())
    message_out.setLocation(
        pyflamegpu.getVariableFloat("x"),
        pyflamegpu.getVariableFloat("y"))
    return pyflamegpu.ALIVE
    
output_func_translated = pyflamegpu.codegen.translate(output_message)
```

Simply becomes:

```py

output_message_str = """
@pyflamegpu.agent_function
def output_message(message_in: pyflamegpu.MessageNone, message_out: pyflamegpu.MessageSpatial2D):
    message_out.setVariableUInt("id", pyflamegpu.getID())
    message_out.setLocation(
        pyflamegpu.getVariableFloat("x"),
        pyflamegpu.getVariableFloat("y"))
    return pyflamegpu.ALIVE
    
output_func_translated = pyflamegpu.codegen.translate(output_message_str)
"""
```

I've created a new issue, https://github.com/FLAMEGPU/FLAMEGPU2/issues/1078, to track this as I'm not confident I'll find a resolution quickly.

**Reply by bytetec** [reply]
yes, without PyCharm it works. Also the workaround works.
![image](https://github.com/FLAMEGPU/FLAMEGPU2/assets/138150718/2320f643-edf8-4fe2-b81a-3ea5349dd44f)

I'm pretty sure that in the 2022 version of PyCharm hadn't this error.
Thanks for opening the issue and specially your help.

**Reply by Robadob** [reply]
Which version of PyCharm are you using? I'll try and reproduce this if I get time this afternoon.

**Reply by bytetec** [reply]
PyCharm 2023.1.2 (Professional Edition)
Build #PY-231.9011.38, built on May 16, 2023

**Reply by Robadob** [reply]
I still can't reproduce this 😢

Both Debug and Run work for me in the latest PyCharm (I couldn't spot a download old version link) and Python 3.11 
```
PyCharm 2023.1.3 (Professional Edition)
Build #PY-231.9161.41, built on June 20, 2023
```

This is all that prints to console in each case, it's visible that `pydevd` is being used.

**Debug**
```
C:\Users\Robert\miniconda3\envs\py311\python.exe "C:/Program Files/JetBrains/PyCharm 2023.1.3/plugins/python/helpers/pydev/pydevd.py" --multiprocess --qt-support=auto --client 127.0.0.1 --port 56357 --file C:\Users\Robert\PycharmProjects\pythonProject\main.py -s 100 
Connected to pydev debugger (build 231.9161.41)

Process finished with exit code 0
```

**Release**
```
C:\Users\Robert\miniconda3\envs\py311\python.exe C:\Users\Robert\PycharmProjects\pythonProject\main.py -s 100 

Process finished with exit code 0
```

Without being able to reproduce the unexpected behaviour, I doubt we can address it unless you want to assist by debugging it on your end.

![image](https://github.com/FLAMEGPU/FLAMEGPU2/assets/742154/1a9721d6-72ef-4ea2-81da-22bf12caa82d)

**Reply by Robadob** [reply]
Managed to install the previous version of PyCharm, same build number as you. Still working, I'm afraid it appears to be a local configuration issue. Any idea what settings you may have changed?

![image](https://github.com/FLAMEGPU/FLAMEGPU2/assets/742154/70fe353e-c996-43af-bf79-ab25e7c56e49)

**Reply by bytetec** [reply]
Very strange, I reinstalled PyCharm and still get the error, but if I debug then it runs through... I will try to reinstall also the python env, let's see if that helps.

**Reply by bytetec** [reply]
Deleted the python environment and created a new one and installed pyflamegpu wheel. Still have the error and debug works....

**Reply by Robadob** [reply]
I tried both Run and Debug within PyCharm and saw no difference.

If you want to pursue this further I would suggest making a copy of the `codegen` subpackage, and using that instead of `pyflamegpu.codegen` (just place the directory in the same directory as your source file, and remove `pyflamegpu.` from `codgen` references in the tutorial code).

From here you can modify `codegen/__init__.py` to print some information that may provide us more information.

I added these two print statements

```py
        module = inspect.getmodule(function)
        print("Module: %s"%(module));
        # get functions
        module_functions = inspect.getmembers(module, inspect.isfunction)
        print("Module Fns:\n%s"%(module_functions));
```

Which printed

```
Module: <module '__main__' from 'C:\\Users\\Robert\\fgpu2\\fgpu2\\examples\\python_native\\tutorial.py'>
Module Fns:
[('input_message', <function input_message at 0x0000024128937600>), ('output_message', <function output_message at 0x00000241268FD9E0>)]
Module: <module '__main__' from 'C:\\Users\\Robert\\fgpu2\\fgpu2\\examples\\python_native\\tutorial.py'>
Module Fns:
[('input_message', <function input_message at 0x0000024128937600>), ('output_message', <function output_message at 0x00000241268FD9E0>)]
```

For you be receiving the error that you are, it would suggest `module_functions` is an empty list, so perhaps the module is different to what's expected.

**Reply by bytetec** [reply]
Exactly, an empty list
`Module: <module '__main__' from 'H:\\example_model2.py'>
Module Fns:
[]`

**Reply by Robadob** [reply]
If you change

```py
module_functions = inspect.getmembers(module, inspect.isfunction)
```

to


```py
module_functions = inspect.getmembers(module)
```

What gets printed? This should be all the symbols from the module e.g. variables, classes, functions.

I get
```
[('AGENT_COUNT', 16384), ('ENV_WIDTH', 25), ('__annotations__', {}), ('__builtins__', <module 'builtins' (built-in)>), ('__cached__', None), ('__doc__', None), ('__file__', 'C:\\Users\\Robadob\\fgpu2\\examples\\python_native\\tutorial.py'), ('__loader__', <_frozen_importlib_external.SourceFileLoader object at 0x000002700F1EA130>), ('__name__', '__main__'), ('__package__', None), ('__spec__', None), ('agent', <pyflamegpu.pyflamegpu.AgentDescription; proxy of <Swig Object of type 'flamegpu::AgentDescription *' at 0x00000270118AEE40> >), ('codegen', <module 'codegen' from 'C:\\Users\\Robadob\\fgpu2\\examples\\python_native\\codegen\\__init__.py'>), ('env', <pyflamegpu.pyflamegpu.EnvironmentDescription; proxy of <Swig Object of type 'flamegpu::EnvironmentDescription *' at 0x0000027011904CC0> >), ('input_message', <function input_message at 0x000002701197FF70>), ('message', <pyflamegpu.pyflamegpu.MessageSpatial2D_Description; proxy of <Swig Object of type 'flamegpu::MessageSpatial2D::Description *' at 0x000002700F79EA50> >), ('model', <pyflamegpu.pyflamegpu.ModelDescription; proxy of <Swig Object of type 'flamegpu::ModelDescription *' at 0x000002700F79EB10> >), ('output_message', <function output_message at 0x000002701197FE50>), ('pyflamegpu', <module 'pyflamegpu.pyflamegpu' from 'C:\\Users\\Robadob\\fgpu2\\build\\lib\\Debug\\python\\venv\\lib\\site-packages\\pyflamegpu\\pyflamegpu.py'>), ('sys', <module 'sys' (built-in)>)]
[('AGENT_COUNT', 16384), ('ENV_WIDTH', 25), ('__annotations__', {}), ('__builtins__', <module 'builtins' (built-in)>), ('__cached__', None), ('__doc__', None), ('__file__', 'C:\\Users\\Robadob\\fgpu2\\examples\\python_native\\tutorial.py'), ('__loader__', <_frozen_importlib_external.SourceFileLoader object at 0x000002700F1EA130>), ('__name__', '__main__'), ('__package__', None), ('__spec__', None), ('agent', <pyflamegpu.pyflamegpu.AgentDescription; proxy of <Swig Object of type 'flamegpu::AgentDescription *' at 0x00000270118AEE40> >), ('codegen', <module 'codegen' from 'C:\\Users\\Robadob\\fgpu2\\examples\\python_native\\codegen\\__init__.py'>), ('env', <pyflamegpu.pyflamegpu.EnvironmentDescription; proxy of <Swig Object of type 'flamegpu::EnvironmentDescription *' at 0x0000027011904CC0> >), ('input_message', <function input_message at 0x000002701197FF70>), ('message', <pyflamegpu.pyflamegpu.MessageSpatial2D_Description; proxy of <Swig Object of type 'flamegpu::MessageSpatial2D::Description *' at 0x000002700F79EA50> >), ('model', <pyflamegpu.pyflamegpu.ModelDescription; proxy of <Swig Object of type 'flamegpu::ModelDescription *' at 0x000002700F79EB10> >), ('output_func_translated', '\n\nFLAMEGPU_AGENT_FUNCTION(output_message, flamegpu::MessageNone, flamegpu::MessageSpatial2D){\n    FLAMEGPU->message_out.setVariable<unsigned int>("id", FLAMEGPU->getID());\n    FLAMEGPU->message_out.setLocation(FLAMEGPU->getVariable<float>("x"), FLAMEGPU->getVariable<float>("y"));\n    return flamegpu::ALIVE;\n}\n'), ('output_message', <function output_message at 0x000002701197FE50>), ('pyflamegpu', <module 'pyflamegpu.pyflamegpu' from 'C:\\Users\\Robadob\\fgpu2\\build\\lib\\Debug\\python\\venv\\lib\\site-packages\\pyflamegpu\\pyflamegpu.py'>), ('sys', <module 'sys' (built-in)>)]
```

Noticeably, the two calls differ, seemingly because the latter call also includes the variable created by the first call.

```
('output_func_translated', '\n\nFLAMEGPU_AGENT_FUNCTION(output_message, flamegpu::MessageNone, flamegpu::MessageSpatial2D){\n    FLAMEGPU->message_out.setVariable<unsigned int>("id", FLAMEGPU->getID());\n    FLAMEGPU->message_out.setLocation(FLAMEGPU->getVariable<float>("x"), FLAMEGPU->getVariable<float>("y"));\n    return flamegpu::ALIVE;\n}\n')
```
Hence, my guess would be for some reason `pyflamegpu.codegen.translate()` is being triggered before the function's have been parsed. The reason why, is unclear. But seeing what has been parsed and the order of your source file may provide a hint.

**Reply by bytetec** [reply]
Sorry writing you back late, I was away.
I tried the change,

Normal run:
```
Module: <module '__main__' from 'H:\\example_model2.py'>
Module Fns: [('__doc__', None), ('__file__', 'H:\\example_model2.py'), ('__loader__', <_frozen_importlib_external.SourceFileLoader object at 0x0000019C55426650>), ('__name__', '__main__'), ('__package__', ''), ('__spec__', None)]
```

PyCharm Debug:
```
Module: <module '__main__' from 'H:\\example_model2.py'>
Module Fns: [('AGENT_COUNT', 16384), ('ENV_WIDTH', 25), ('__builtins__', <module 'builtins' (built-in)>), ('__doc__', None), ('__file__', 'H:\\example_model2.py'), ('__loader__', <_frozen_importlib_external.SourceFileLoader object at 0x00000290E16168D0>), ('__name__', '__main__'), ('__package__', ''), ('__spec__', None), ('agent', <pyflamegpu.pyflamegpu.AgentDescription; proxy of <Swig Object of type 'flamegpu::AgentDescription *' at 0x00000290E4EDAA30> >), ('codegen', <module 'codegen' from 'H:\\codegen\\__init__.py'>), ('env', <pyflamegpu.pyflamegpu.EnvironmentDescription; proxy of <Swig Object of type 'flamegpu::EnvironmentDescription *' at 0x00000290E4EDAAC0> >), ('input_message', <function input_message at 0x00000290E4F3C360>), ('math', <module 'math' (built-in)>), ('message', <pyflamegpu.pyflamegpu.MessageSpatial2D_Description; proxy of <Swig Object of type 'flamegpu::MessageSpatial2D::Description *' at 0x00000290E4EDA910> >), ('model', <pyflamegpu.pyflamegpu.ModelDescription; proxy of <Swig Object of type 'flamegpu::ModelDescription *' at 0x00000290E4EDA880> >), ('output_func_translated', '\n\nFLAMEGPU_AGENT_FUNCTION(output_message, flamegpu::MessageNone, flamegpu::MessageSpatial2D){\n    FLAMEGPU->message_out.setVariable<unsigned int>("id", FLAMEGPU->getID());\n    FLAMEGPU->message_out.setLocation(FLAMEGPU->getVariable<float>("x"), FLAMEGPU->getVariable<float>("y"));\n    return flamegpu::ALIVE;\n}\n'), ('output_message', <function output_message at 0x00000290E4AEBCE0>), ('pyflamegpu', <module 'pyflamegpu.pyflamegpu' from 'H:\\python_envs\\python3.11\\Lib\\site-packages\\pyflamegpu\\pyflamegpu.py'>), ('sys', <module 'sys' (built-in)>)]
``` 
In debug mode I'm missing the __annotations__ part.

**Reply by Robadob** [reply]
Thanks, I'll try and look into this more later in the week I'm working on a different project today.

**Reply by Robadob** [reply]
I'm no closer to understanding this unusual behaviour. A couple more tests to help narrow it down.

-------------

**Perform the inspect within `example_model2.py`** (this can narrow down whether its `inspect` in general failing, or the process of passing the function)

e.g.
```
from pyflamegpu import *
import inspect

@pyflamegpu.agent_function
def output_message(message_in: pyflamegpu.MessageNone, message_out: pyflamegpu.MessageSpatial2D):
    message_out.setVariableUInt("id", pyflamegpu.getID())
    message_out.setLocation(
        pyflamegpu.getVariableFloat("x"),
        pyflamegpu.getVariableFloat("y"))
    return pyflamegpu.ALIVE

print(inspect.getmodule(output_message))
````

I would expect this to output the same as it did when you printed the result of `inspect.getmodule()` inside `codegen/__init__.py`.

---------------------------------

**Check whether it's actually being see as a Python (or even C) function.** Maybe an aggressive implementation is optimising it?

Simply modify `codegen/__init__.py` (or the main file if that fails the previous test too), to include

```
        print("isroutine %s"%(inspect.isroutine(function))) #  C or Py function
        print("isfunction %s"%(inspect.isfunction(function))) # Py function
```

Both of these output as True for me, not really sure why they wouldn't. But I'm trying to understand enough so that I can post a question on StackOverflow at this point, given the Python docs for [`inspect`](https://docs.python.org/3/library/inspect.html) don't contain any obvious hints to your problem.

------------

*Without the ability to reproduce locally, it's difficult for me to have a significant play to narrow it down. Likewise, I'm not too knowledgeable on Python's internals.*

**Reply by bytetec** [reply]
I found out that I had a different run execution configuration in PyCharm. If configured to run in Python Console, it fails. Switching to Emulate Terminal solved the problem.
![image](https://github.com/FLAMEGPU/FLAMEGPU2/assets/138150718/3e259c0b-0a22-4ad1-9d54-66e6fc08d8a4)
To analysis the problem I also tried the code in VSC and I got even another result as the program does only one step. No error, just one "100.00 Drift correct" 
Looks like I'm fighting more with the tools than with the code...
Thank you for helping me and your time investment.

**Reply by Robadob** [reply]
I'll try and read up on that feature, and see what the difference is incase we can find a way to workaround it within the codegen module.

> I got even another result as the program does only one step. No error, just one "100.00 Drift correct"

One step is the new(?) default. It would previously run forever by default, if steps was not specified. It was decided this was undesirable for non-visualisation runs. If you pass `-s 0` at runtime, you will get an endless run. Otherwise you can call `cudaModel.SimulationConfig().steps=0` to programmatically set the desired number of steps.

"Drift correct" is a step function attached to many of the circles models, it was our attempt at adding a non-visualisation validation. It's not required and can be removed from the model if preferred.

Please let me know if you think the documentation for the tutorial could clarify this better, It's covered [here](https://docs.flamegpu.com/tutorial/index.html#running-the-simulation).

Thanks for all your involvement in debugging this too, really helpful to have someone so happy to help debug the problem.

**Reply by bytetec** [reply]
I would have some proposals for enhancing and simplify the documentation. Should I reuse this thread or open a new one?

**Reply by Robadob** [reply]
Feel free to create an issue on the docs repo, will make sure it gets seen.

https://github.com/FLAMEGPU/FLAMEGPU2-docs/issues
